airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/hacks.py

```diff
@@ -44,48 +31,56 @@ def _check_for_invalid_write_path(write_path: str):
         )
 
 
-def write_registry_to_overrode_file_paths(
+def construct_registry_entry_write_path(
     registry_entry: PolymorphicRegistryEntry,
     registry_name: str,
-    metadata_entry: LatestMetadataEntry,
-    registry_directory_manager: GCSFileManager,
-) -> GCSFileHandle:
+) -> str:
     """
-    Write the registry entry to the docker repository and version specific file paths
-    in the event that the docker repository is overridden.
+    Construct a registry entry write path from its parts.
 
     Underlying issue:
-        The registry entry files (oss.json and cloud.json) are traditionally written to
-        the same path as the metadata.yaml file that created them. This is fine for the
-        most cases, but when the docker repository is overridden, the registry entry
-        files need to be written to a different path.
+        This is barely a hack.
+
+        But it is related to a few imperfect design decisions that we have to work around.
+        1. Metadata files and the registry entries are saved to the same top level folder.
+        2. That save path is determined by the docker repository and version of the image
+        3. A metadata file can include overrides for the docker repository and version of the image depending on the registry
+        4. The platform looks up registry entries by docker repository and version of the image.
+        5. The registry generation depends on what ever registry entry is written to a path ending in latest/{registry_name}.json
+
+        This means that when a metadata file overrides the docker repository and version of the image,
+        the registry entry needs to be written to a different path than the metadata file.
+
+        *But only in the case that its a versioned path and NOT a latest path.*
+
+        Example:
+        If metadata file for source-posgres is at version 2.0.0 but there is a override for the cloud registry
+        that changes the docker repository to source-postgres-strict-encrypt and the version to 1.0.0
+
+        Then we will have a metadata file written to:
+        gs://my-bucket/metadata/source-postgres/2.0.0/metadata.yaml
+
+        and registry entries written to:
+        gs://my-bucket/metadata/source-postgres/2.0.0/oss.json
+        gs://my-bucket/metadata/source-postgres-strict-encrypt/1.0.0/cloud.json
```

Comment:
Why do we need to rewrite the registry entry for the old version? Should we _only_ modify `latest`?

My thought here is that if when we published 1.0.0 there were some different fields like suggested streams applicable for 1.0.0, now we would be grabbing the suggested streams from 2.0.0 and using those instead - is that right?

My feeling is that we should leave the old one be and only tweak latest when dealing with overrides, but curious to know what your thoughts are here and if there's some benefit I'm missing.

Reply:
I think if i had a magic wand the ideal world-view would be to separate this override stuff from rollbacks/version changes. Reason being that if I want to now use 1.0.0, I really do want all the versioned metadata that was present in 1.0.0 to be used, not just some pieces of it (e.g. spec, image tag).

Reply:
@pedroslopez Good news is I think you, me and Ella all agree on that ideal world view.

So heres the contextual facts from Ella and I's chat last week :
1. We only have overrides today
2. That leads to all configuration settings that affect our Registries live in the master branch
3. This is useful if you consider a case that the platform requires a new metadata field, we dont have to do any special backfilling of old registry entries / metadata files for overrode docker images
4. However this is not useful (specifically for rollbacks) as you have to be aware of if an overrode docker image had any metadata requirements 
5. and to add to this, Launch Darkly Overides behave differently to Registry Overrides. With launch darkly you can time travel as we reference the versioned registry entry directly.
6. LD refs combined with Metadata overrides (in the current state) lead to a hole where a LD override cannot find a Registry Entry for a strict-encrypt version.

What this leads us to believe is:
1. We need BOTH the ability to OVERRIDE metadata per registry and the ability to REF a different version of the metadata for the registry.
2. If refs become the prefered method then LD and Registry overrides dont deviate in behaviour which is a good thing
2. But! If we do introduce REFs this will require us to think hard about required registry fields, out of date refs (and how to update them), and global vs versioned metadata (tech spec territory)
3. And if were in tech spec territory then we also have the ability to revisit "Should we be using docker image information as our source of truth for versioning and paths"

**TLDR: We should do refs, but we need to explicitly design it, in the meantime we should plug this hole in our paths that LD references**

Reply:
Good question @pedroslopez :

So the main culprit here is a few facts
1. Connectors are a combination of artifacts (dockerfiles) and metadata
1. We use docker image tags as connector versions
1. Metadata can change without a new docker image tag being required
1. There are two instances of a connector (cloud and oss) per version

This means that anytime we update the metadata for a connector, we need to apply that new metadata to all "latest" versions for cloud and oss.

For example
1. if the metadata for source-s3 is set so that oss get v2 and cloud gets v1
2. then platform says having field `xyz` is now mandatory
3. we need to update the metadata to include xyz so that the registry entry for v2 and the registry entry for v1 have xyz

Happy to jump on a call if that helps 😅

Reply:
> 3. we need to update the metadata to include xyz so that the registry entry for v2 and the registry entry for v1 have xyz

@bnchrch I think this is the bit I don't quite understand - why does v1 now need xyz if we're not copying all of v1 metadata to latest when an override is in place but rather just tweaking the docker_image_tag field of whatever's latest?

Feels like we're in between two states:
1. The current world where we have a set of metadata and just override the docker_image_tag of that metadata
2. The ideal world where saying we want to use v1 means we use v1's metadata specific to that version

If we're in world view 1 and the definition of specifying an override is "use this other value for this field instead", then I don't think we should be doing anything with the existing data of other versions, but just make latest match with whatever override you set on these specific fields.

Does that make sense? Also happy to jump on a call